android/src/main/java/com/expensify/livemarkdown/MarkdownUtils.java

```diff
@@ -7,78 +7,65 @@
 import androidx.annotation.NonNull;
 
 import com.expensify.livemarkdown.spans.*;
+import com.facebook.react.bridge.ReactContext;
+import com.facebook.react.util.RNLog;
 import com.facebook.react.views.text.internal.span.CustomLineHeightSpan;
 import com.facebook.soloader.SoLoader;
 
 import org.json.JSONArray;
 import org.json.JSONException;
 import org.json.JSONObject;
 
-import java.io.IOException;
-import java.io.InputStream;
 import java.util.Objects;
 
 public class MarkdownUtils {
   static {
     SoLoader.loadLibrary("livemarkdown");
   }
 
-  private static boolean IS_RUNTIME_INITIALIZED = false;
+  private static synchronized native String nativeParseMarkdown(String input, int parserId);
 
-  public static synchronized void maybeInitializeRuntime(AssetManager assetManager) {
-    if (IS_RUNTIME_INITIALIZED) {
-      return;
-    }
-    try {
-      InputStream inputStream = assetManager.open("react-native-live-markdown-parser.js");
-      byte[] buffer = new byte[inputStream.available()];
-      inputStream.read(buffer);
-      inputStream.close();
-      String code = new String(buffer);
-      nativeInitializeRuntime(code);
-      IS_RUNTIME_INITIALIZED = true;
-    } catch (IOException e) {
-      throw new RuntimeException("Failed to initialize Markdown runtime");
-    }
-  }
-
-  private static native void nativeInitializeRuntime(String code);
-
-  private synchronized static String parseMarkdown(String input) {
-    return nativeParseMarkdown(input);
-  }
-
-  private static native String nativeParseMarkdown(String input);
-
-  public MarkdownUtils(@NonNull AssetManager assetManager) {
-    mAssetManager = assetManager;
+  public MarkdownUtils(@NonNull ReactContext reactContext) {
+    mReactContext = reactContext;
+    mAssetManager = reactContext.getAssets();
   }
 
+  private final @NonNull ReactContext mReactContext;
   private final @NonNull AssetManager mAssetManager;
 
   private String mPrevInput;
-
   private String mPrevOutput;
+  private int mPrevParserId;
 
   private MarkdownStyle mMarkdownStyle;
+  private int mParserId;
 
   public void setMarkdownStyle(@NonNull MarkdownStyle markdownStyle) {
     mMarkdownStyle = markdownStyle;
   }
 
+  public void setParserId(int parserId) {
+    mParserId = parserId;
+  }
+
   public void applyMarkdownFormatting(SpannableStringBuilder ssb) {
     Objects.requireNonNull(mMarkdownStyle, "mMarkdownStyle is null");
 
     removeSpans(ssb);
 
     String input = ssb.toString();
     String output;
-    if (input.equals(mPrevInput)) {
+    if (input.equals(mPrevInput) && mParserId == mPrevParserId) {
       output = mPrevOutput;
     } else {
-      output = parseMarkdown(input);
+      try {
+        output = nativeParseMarkdown(input, mParserId);
+      } catch (Exception e) {
+        output = "[]";
+      }
```

Comment:
`nativeParseMarkdown` calls `WorkletRuntime::runGuarded` so the error is presented to the user in RedBox. This is just to prevent a native crash. I'd like to improve error handling in a separate PR since this part of the library needs a small refactor and I don't want to spoil this PR with too many changes.